include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,454 @@
+    {}
+
+    template <typename FactoryParameters>
+    explicit EnableBatchSolver(std::shared_ptr<const Executor> exec,
+                               std::shared_ptr<const BatchLinOp> system_matrix,
+                               const FactoryParameters& params)
+        : BatchSolver(system_matrix, nullptr, params.tolerance,
+                      params.max_iterations, params.tolerance_type),
+          EnableBatchLinOp<ConcreteSolver, PolymorphicBase>(
+              exec, gko::transpose(system_matrix->get_size()))
+    {
+        GKO_ASSERT_BATCH_HAS_SQUARE_DIMENSIONS(system_matrix_);
+
+        using value_type = typename ConcreteSolver::value_type;
+        using Identity = matrix::Identity<value_type>;
+        using real_type = remove_complex<value_type>;
+
+        if (params.generated_preconditioner) {
+            GKO_ASSERT_BATCH_EQUAL_DIMENSIONS(params.generated_preconditioner,
+                                              this);
+            preconditioner_ = std::move(params.generated_preconditioner);
+        } else if (params.preconditioner) {
+            preconditioner_ = params.preconditioner->generate(system_matrix_);
+        } else {
+            auto id = Identity::create(exec, system_matrix->get_size());
+            preconditioner_ = std::move(id);
+        }
+        const size_type workspace_size = system_matrix->get_num_batch_items() *
+                                         (sizeof(real_type) + sizeof(int));
+        workspace_.set_executor(exec);
+        workspace_.resize_and_reset(workspace_size);
+    }
+
+    void set_system_matrix(std::shared_ptr<const BatchLinOp> new_system_matrix)
+    {
+        auto exec = self()->get_executor();
+        if (new_system_matrix) {
+            GKO_ASSERT_BATCH_EQUAL_DIMENSIONS(self(), new_system_matrix);
+            GKO_ASSERT_BATCH_HAS_SQUARE_DIMENSIONS(new_system_matrix);
+            if (new_system_matrix->get_executor() != exec) {
+                new_system_matrix = gko::clone(exec, new_system_matrix);
+            }
+        }
+        this->set_system_matrix_base(new_system_matrix);
+    }
+
+    /**
+     * Creates a shallow copy of the provided system matrix, clones it onto
+     * this executor if executors don't match.
+     */
+    EnableBatchSolver& operator=(const EnableBatchSolver& other)
+    {
+        if (&other != this) {
+            this->set_size(other.get_size());
+            set_system_matrix(other.get_system_matrix());
+        }
+        return *this;
+    }
+
+    /**
+     * Moves the provided system matrix, clones it onto this executor if
+     * executors don't match. The moved-from object has a nullptr system matrix.
+     */
+    EnableBatchSolver& operator=(EnableBatchSolver&& other)
+    {
+        if (&other != this) {
+            this->set_size(other.get_size());
+            set_system_matrix(other.get_system_matrix());
+            other.set_system_matrix(nullptr);
```

Comment:
The core solvers will also copy/move the preconditioner. That should be mirrored here.